src/Razor/test/RazorLanguage.Test/TestFiles/IntegrationTests/ComponentDesignTimeCodeGenerationTest/GenericComponent_GenericEventCallback_TypeInference/TestComponent.codegen.cs

```diff
@@ -0,0 +1,72 @@
+// <auto-generated/>
+#pragma warning disable 1591
+namespace Test
+{
+    #line hidden
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using Microsoft.AspNetCore.Components;
+#nullable restore
+#line 1 "x:\dir\subdir\Test\TestComponent.cshtml"
+using Test;
+
+#line default
+#line hidden
+#nullable disable
+    public class TestComponent : Microsoft.AspNetCore.Components.ComponentBase
+    {
+        #pragma warning disable 219
+        private void __RazorDirectiveTokenHelpers__() {
+        }
+        #pragma warning restore 219
+        #pragma warning disable 0414
+        private static System.Object __o = null;
+        #pragma warning restore 0414
+        #pragma warning disable 1998
+        protected override void BuildRenderTree(Microsoft.AspNetCore.Components.Rendering.RenderTreeBuilder builder)
+        {
+            __Blazor.Test.TestComponent.TypeInference.CreateMyComponent_0(builder, -1, -1, 
+#nullable restore
+#line 2 "x:\dir\subdir\Test\TestComponent.cshtml"
+                   3
+
+#line default
+#line hidden
+#nullable disable
+            , -1, Microsoft.AspNetCore.Components.EventCallback.Factory.Create<Test.MyEventArgs>(this, 
+#nullable restore
+#line 2 "x:\dir\subdir\Test\TestComponent.cshtml"
+                               x => {}
+
+#line default
+#line hidden
+#nullable disable
+            ));
+#nullable restore
+#line 2 "x:\dir\subdir\Test\TestComponent.cshtml"
+__o = typeof(MyComponent<>);
+
+#line default
+#line hidden
+#nullable disable
+        }
+        #pragma warning restore 1998
+    }
+}
+namespace __Blazor.Test.TestComponent
+{
+    #line hidden
+    internal static class TypeInference
+    {
+        public static void CreateMyComponent_0<TItem>(global::Microsoft.AspNetCore.Components.Rendering.RenderTreeBuilder builder, int seq, int __seq0, TItem __arg0, int __seq1, global::Microsoft.AspNetCore.Components.EventCallback<global::Test.MyEventArgs> __arg1)
```

Comment:
Prior to this change, we would generate

`global::Microsoft.AspNetCore.Components.EventCallback<Test.MyEventArgs> __arg1)`

which would fail to compile when the compiler attempts to look up the type under the `__Blazor.Test` namespace.

We intentionally do not globally qualify tag helper descriptors when we discover them which "might" have avoided doing this mangling, but presumably changing that has significantly larger ramifications.

Reply:
> We intentionally do not globally qualify tag helper descriptors when we discover them which "might" have avoided doing this mangling, but presumably changing that has significantly larger ramifications.

This is a super important concept. We use metadata name concept as the **canonical representation** of type names. When we need to generate code, we often global-qualify it because that's what's needed in the generated code.

If we used global-qualified names as the **canonical representation** we'd have code in lots of places to *unqualify* it for display to users.